pkg/sampling/oteltracestate_test.go

```diff
@@ -0,0 +1,336 @@
+		{"extra:stuff;rv:22222222222222", "22222222222222", ns, []string{"extra:stuff"}, nil},
+		{"rv:ffffffffffffff", "ffffffffffffff", ns, nil, nil},
+		{"rv:88888888888888", "88888888888888", ns, nil, nil},
+		{"rv:00000000000000", "00000000000000", ns, nil, nil},
+
+		// r-value range error (15 bytes of hex or more)
+		{"rv:100000000000000", ns, ns, nil, ErrRValueSize},
+		{"rv:fffffffffffffffff", ns, ns, nil, ErrRValueSize},
+
+		// no trailing ;
+		{"x:1;", ns, ns, nil, strconv.ErrSyntax},
+
+		// empty key
+		{"x:", ns, ns, []string{"x:"}, nil},
+
+		// charset test
+		{"x:0X1FFF;y:.-_-.;z:", ns, ns, []string{"x:0X1FFF", "y:.-_-.", "z:"}, nil},
+		{"x1y2z3:1-2-3;y1:y_1;xy:-;th:50", ns, "50", []string{"x1y2z3:1-2-3", "y1:y_1", "xy:-"}, nil},
+
+		// size exceeded
+		{"x:" + strings.Repeat("_", 255), ns, ns, nil, ErrTraceStateSize},
+		{"x:" + strings.Repeat("_", 254), ns, ns, []string{"x:" + strings.Repeat("_", 254)}, nil},
+	} {
+		t.Run(testName(test.in), func(t *testing.T) {
+			otts, err := NewOTelTraceState(test.in)
+
+			if test.expectErr != nil {
+				require.True(t, errors.Is(err, test.expectErr), "%q: not expecting %v wanted %v", test.in, err, test.expectErr)
+			} else {
+				require.NoError(t, err)
+			}
+			if test.rval != ns {
+				require.True(t, otts.HasRValue())
+				require.Equal(t, test.rval, otts.RValue())
+			} else {
+				require.False(t, otts.HasRValue(), "should have no r-value: %s", otts.RValue())
+			}
+			if test.tval != ns {
+				require.True(t, otts.HasTValue())
+				require.Equal(t, test.tval, otts.TValue())
+			} else {
+				require.False(t, otts.HasTValue(), "should have no t-value: %s", otts.TValue())
+			}
+			var expect []KV
+			for _, ex := range test.extra {
+				k, v, _ := strings.Cut(ex, ":")
+				expect = append(expect, KV{
+					Key:   k,
+					Value: v,
+				})
+			}
+			require.Equal(t, expect, otts.ExtraValues())
+
+			if test.expectErr != nil {
+				return
+			}
+			// on success Serialize() should not modify
+			// test by re-parsing
+			var w strings.Builder
+			require.NoError(t, otts.Serialize(&w))
+			cpy, err := NewOTelTraceState(w.String())
+			require.NoError(t, err)
+			require.Equal(t, otts, cpy)
+		})
+	}
+}
+
+func TestUpdateTValueWithSampling(t *testing.T) {
+	type testCase struct {
+		// The input otel tracestate; no error conditions tested
+		in string
+
+		// The incoming adjusted count; defined whether
+		// t-value is present or not.
+		adjCountIn float64
+
+		// the update probability; threshold and tvalue are
+		// derived from this
+		prob float64
+
+		// when update error is expected
+		updateErr error
+
+		// output t-value
+		out string
+
+		// output adjusted count
+		adjCountOut float64
+	}
+	for _, test := range []testCase{
+		// 8/16 in, sampled at (0x10-0xe)/0x10 = 2/16 => adjCount 8
+		{"th:8", 2, 0x2p-4, nil, "th:e", 8},
+
+		// 8/16 in, sampled at 14/16 => no update, adjCount 2
+		{"th:8", 2, 0xep-4, nil, "th:8", 2},
```

Comment:
I don't understand this case, shouldn't this be an `ErrInconsistentSampling`? It's raising the sampling probability from 50% to 87.5%, isn't it?

Reply:
Nice catch! You are correct and the test was not checking error cases properly. This test input did create `ErrInconsistentSampling`, the test and this specific case [have been fixed](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/29720/commits/e294124487562c9109d31244bc1f22c7e24a525c).